mat64/vector.go

```diff
@@ -357,6 +360,61 @@ func (v *Vector) MulVec(a Matrix, b *Vector) {
 	}
 }
 
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Vector is little-endian encoded as follows:
+//   0 -  8  number of elements     (int64)
+//   8 - ..  vector's data elements (float64)
+func (v Vector) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, sizeInt64+v.n*sizeFloat64))
+	err := binary.Write(buf, defaultEndian, int64(v.n))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, val := range v.mat.Data {
+		err = binary.Write(buf, defaultEndian, val)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes the binary form into the receiver.
+// It panics if the receiver is a non-zero Vector.
+//
+// See MarshalBinary for the on-disk layout.
+func (v *Vector) UnmarshalBinary(data []byte) error {
+	if v.n != 0 || len(v.mat.Data) != 0 {
+		panic("mat64: unmarshal into non-zero vector")
+	}
+
+	buf := bytes.NewReader(data)
+	var n int64
+	err := binary.Read(buf, defaultEndian, &n)
+	if err != nil {
+		return err
+	}
+
+	if n < 0 || n > maxInt {
+		return matrix.ErrShape
+	}
+
+	v.n = int(n)
+	v.mat.Inc = 1
+	v.mat.Data = use(v.mat.Data, v.n)
+
+	for i := range v.mat.Data {
+		err = binary.Read(buf, defaultEndian, &v.mat.Data[i])
```

Comment:
The third argument could be `v.mat.Data` directly. The documentation says `... or a slice of fixed-size values`.